adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -136,35 +140,127 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
     replaceRdd(tFn(rdd))
   }
 
+  // The partition map is structured as follows:
+  // The outer option is for whether or not there is a partition map.
+  //   - This is None in the case that we don't know the bounds on each 
+  //     partition.
+  // The Array is the length of the number of partitions.
+  // The inner option is in case there is no data on a partition.
+  // The (ReferenceRegion, ReferenceRegion) tuple contains the bounds of the 
+  //   partition, such that the lowest start is first and the highest end is
+  //   second.
+  protected val optPartitionMap: Option[Array[Option[(ReferenceRegion, ReferenceRegion)]]]
+
+  assert(optPartitionMap.isEmpty ||
+    optPartitionMap.exists(_.length == rdd.partitions.length),
+    "Partition Map length differs from number of partitions.")
+
+  val isSorted: Boolean = optPartitionMap.isDefined
+
+  /**
+   * Repartitions all data in rdd and distributes it as evenly as possible
+   * into the number of partitions provided.
+   *
+   * @param partitions the number of partitions to repartition this rdd into
+   * @return a new repartitioned GenomicRDD
+   */
+  private[rdd] def evenlyRepartition(partitions: Int)(implicit tTag: ClassTag[T]): U = {
+    require(isSorted, "Cannot evenly repartition an unsorted RDD.")
+    val count = rdd.count
+    // we don't want a bunch of empty partitions, so we will just use count in
+    // the case the user wants more partitions than rdd records.
+    val finalPartitionNumber = min(count, partitions)
+    // the average number of records on each node will help us evenly repartition
+    val average = count.toDouble / finalPartitionNumber
+
+    val finalPartitionedRDD =
+      flattenRddByRegions()
+        .zipWithIndex
+        .mapPartitions(iter => {
+          // divide the global index by the average to get the destination
+          // partition number
+          iter.map(_.swap).map(f =>
+            ((f._2._1, (f._1 / average).toInt), f._2._2))
+        }, preservesPartitioning = true)
+        .repartitionAndSortWithinPartitions(
+          ManualRegionPartitioner(finalPartitionNumber.toInt))
+
+    finalPartitionedRDD.cache()
+
+    val newPartitionMap = finalPartitionedRDD.mapPartitions(iter =>
+      getRegionBoundsFromPartition(
+        iter.map(f => (f._1._1, f._2))),
+      preservesPartitioning = true).collect
+
+    replaceRdd(finalPartitionedRDD.values, Some(newPartitionMap))
+  }
+
+  /**
+   * If the specified validation strategy is STRICT, throw an exception,
+   * if LENIENT, log a warning, otherwise does nothing.
+   *
+   * @param message The error or warning message.
+   * @param stringency The validation stringency.
+   * @throws IllegalArgumentException If stringency is STRICT.
```

Comment:
Nit: add whitespace separating `@throws` from `@param`. Also, I think we typically have `@throws` before `@param`.